### PR TITLE
overlord/snapstate, snap: support base = "none"

### PR DESCRIPTION
**Pull request by stolowski**
Allow base: none and don't attempt to install it as part of prerequisites. Reject snaps with "base: none" if they have apps or hooks.

### CODE REVIEW

File: cmd/snap/cmd_run.go [hunk 1]
```diff
@@ -897,6 +897,10 @@ func (x *cmdRun) runSnapConfine(info *snap.Info, securityTag, snapApp, hook stri
 	if info.NeedsClassic() {
 		cmd = append(cmd, "--classic")
 	}
+	// XXX: should we prevent any hooks/apps when validating snaps with "base: none"?
```
Comment (zyga) [inline cmd/snap/cmd_run.go]: IMO yes, we should.

File: image/image.go [hunk 2]
```diff
@@ -412,7 +412,7 @@ func neededDefaultProviders(info *snap.Info) (cps []string) {
 // snaps. If not an error is returned.
 func hasBase(snap *snap.Info, local *localInfos, snaps []string) error {
 	// snap needs no base: nothing to do
-	if snap.Base == "" {
+	if snap.Base == "" || snap.Base == "none" {
```
Comment (zyga) [inline image/image.go]: Is `snap.Base` being empty the same as `snap.Base` being equal to `none`? Is using `core` always explicit?

Comment (pedronis) [reply image/image.go]: no,  we never see "core",  core becomes empty (so there is some ambiguity left)

Comment (zyga) [reply image/image.go]: @stolowski I think this warrants a comment.

File: cmd/snap/cmd_run.go [hunk 3]
```diff
@@ -897,6 +897,11 @@ func (x *cmdRun) runSnapConfine(info *snap.Info, securityTag, snapApp, hook stri
 	if info.NeedsClassic() {
 		cmd = append(cmd, "--classic")
 	}
+
+	// this should never happen since we validate snaps with "base: none" and do not allow hooks/apps
+	if info.Base == "none" {
+		return fmt.Errorf(`cannot run hooks / applications with base type "none"`)
```
Comment (bboozzoo) [inline cmd/snap/cmd_run.go]: This probably needs a unit test.

File: tests/main/base-none/task.yaml [hunk 4]
```diff
@@ -0,0 +1,13 @@
+summary: Check that a snap with base "none" can be installed
+
+execute: |
+    #shellcheck source=tests/lib/snaps.sh
+    . "$TESTSLIB"/snaps.sh
+
+    echo "Install a snap with 'base: none'"
+
+    snap pack "$TESTSLIB/snaps/test-snapd-base-none"
+    snap install --dangerous test-snapd-base-none_1.0_all.snap
```
Comment (bboozzoo) [inline tests/main/base-none/task.yaml]: Should probably check a scenario in which we try to install a snap with base: none and some apps.

File: overlord/snapstate/check_snap.go [hunk 5]
```diff
@@ -351,6 +351,12 @@ func checkBases(st *state.State, snapInfo, curInfo *snap.Info, flags Flags, devi
 	if snapInfo.Base == "" {
 		return nil
 	}
+	if snapInfo.Base == "none" {
+		if len(snapInfo.Hooks) > 0 || len(snapInfo.Apps) > 0 {
+			return fmt.Errorf("base type %q not supported with apps or hooks", snapInfo.Base)
```
Comment (bboozzoo) [inline overlord/snapstate/check_snap.go]: Should there be a similar check in snap validation code, so that snap pack complains about it too?

File: tests/main/base-none/task.yaml (hunk 4 above)
Comment (stolowski) [reply tests/main/base-none/task.yaml]: Added.

File: overlord/snapstate/check_snap.go (hunk 5 above)
Comment (stolowski) [reply overlord/snapstate/check_snap.go]: Fair point, I think validation is actually the correct place, then here we only need a simple check ("none" -> return nil). Done.

File: cmd/snap/cmd_run.go (hunk 3 above)
Comment (stolowski) [reply cmd/snap/cmd_run.go]: Added.

File: image/image.go (hunk 2 above)
Comment (stolowski) [reply image/image.go]: Added.

File: cmd/snap/cmd_run.go (hunk 1 above)
Comment (stolowski) [reply cmd/snap/cmd_run.go]: Done.

File: cmd/snap/cmd_run_test.go [hunk 6]
```diff
@@ -712,11 +745,11 @@ func (s *RunSuite) TestAntialiasHappy(c *check.C) {
 	app, outArgs = snaprun.Antialias("alias", inArgs)
 	c.Check(app, check.Equals, "an-app")
 	c.Check(outArgs, check.DeepEquals, []string{
-		"99", // COMP_TYPE (no change)
-		"99", // COMP_KEY (no change)
-		"11", // COMP_POINT (+1 because "an-app" is one longer than "alias")
-		"2",  // COMP_CWORD (no change)
-		" ",  // COMP_WORDBREAKS (no change)
+		"99",                    // COMP_TYPE (no change)
```
Comment (pedronis) [inline cmd/snap/cmd_run_test.go]: is this a gofmt change? it seems unrelated

File: image/image.go [hunk 7]
```diff
@@ -411,10 +411,16 @@ func neededDefaultProviders(info *snap.Info) (cps []string) {
...
 // snaps. If not an error is returned.
 func hasBase(snap *snap.Info, local *localInfos, snaps []string) error {
-	// snap needs no base: nothing to do
+	// snap needs no base (or it simply needs core which is never listed explicitly): nothing to do
 	if snap.Base == "" {
 		return nil
 	}
+
+	// snap explicitly listed as not needing a base snap (e.g. a content-only snap)
+	if snap.Base == "none" {
```
Comment (pedronis) [inline image/image.go]: this merits a test

File: snap/validate.go [hunk 8]
```diff
@@ -356,6 +354,21 @@ func Validate(info *Info) error {
 	return ValidateLayoutAll(info)
 }
 
+// ValidateBase validates the base field.
+func ValidateBase(info *Info) error {
+	// validate that bases do not have base fields
+	if info.Type == TypeOS || info.Type == TypeBase {
+		if info.Base != "" {
```
Comment (pedronis) [inline snap/validate.go]: Base can also be "none" for these now, that is also fine

File: cmd/snap/cmd_run_test.go (hunk 6 above)
Comment (bboozzoo) [reply cmd/snap/cmd_run_test.go]: Quite possible, 1.11 changed the maximum blank fill length between a value and a comment. I'm quite sure this will break our gofmt checks on travis too.

Comment (bboozzoo) [reply cmd/snap/cmd_run_test.go]: That's the change: https://golang.org/doc/go1.11#gofmt

Comment (stolowski) [reply cmd/snap/cmd_run_test.go]: Yep, go fmt, I was actually running it with go 1.12.x and indeed it made travis fail with 1.10. Going to reformat with 1.10.

File: image/image.go (hunk 7 above)
Comment (stolowski) [reply image/image.go]: Added.

File: snap/validate.go (hunk 8 above)
Comment (stolowski) [reply snap/validate.go]: Done.

File: snap/validate.go [hunk 9]
```diff
@@ -356,6 +354,21 @@ func Validate(info *Info) error {
...
+// ValidateBase validates the base field.
+func ValidateBase(info *Info) error {
+	// validate that bases do not have base fields
+	if info.Type == TypeOS || info.Type == TypeBase {
+		if info.Base != "" && info.Base != "none" {
+			return fmt.Errorf(`cannot have "base" field on %q snap %q`, info.Type, info.InstanceName())
+		}
+	}
+
+	if info.Base == "none" && (len(info.Hooks) > 0 || len(info.Apps) > 0) {
```
Comment (zyga) [inline snap/validate.go]: At this stage, do we know about implicitly defined hooks?

Comment (pedronis) [reply snap/validate.go]: the main caller in info.go should have found those

File: tests/main/base-none/task.yaml [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+    snap list | MATCH test-snapd-base-none
+
+    echo "Snap with base: none and apps is rejected"
+    snap pack "$TESTSLIB/snaps/test-snapd-base-none-invalid" 2>&1 | MATCH  "cannot pack .* cannot validate snap \"test-snapd-base-none\".* cannot have apps or hooks with base \"none\".*"
+
+    echo "Creating and installing an invalid snap"
+    # we cannot use snap pack, create snap manually
+    mksquashfs "$TESTSLIB/snaps/test-snapd-base-none-invalid" test-snapd-base-none-invalid_x1.snap -comp xz -no-fragments -no-progress
+    snap install --dangerous test-snapd-base-none-invalid_x1.snap 2>&1 | MATCH "error: cannot read snap file: cannot have apps or hooks with base \"none\""
+
```
Comment (bboozzoo) [inline tests/main/base-none/task.yaml]: Stray indent without newline? Fix only if you feel it's worth doing another spread run.

File: tests/lib/snaps/test-snapd-base-none-invalid/meta/snap.yaml [hunk 11]
```diff
@@ -0,0 +1,8 @@
+name: test-snapd-base-none
+base: none
+version: 1.0
+summary: A test snap that uses base "none"
+
+apps:
+    cmd1:
+        command: bin/cmd
```
Comment (bboozzoo) [inline tests/lib/snaps/test-snapd-base-none-invalid/meta/snap.yaml]: Missing newline

File: tests/main/base-none/task.yaml (hunk 10 above)
Comment (stolowski) [reply tests/main/base-none/task.yaml]: Uhm. Yep I'll land it as is while world is green. Thank you.
